src/app/gps/qgsgpstoolbar.cpp

```diff
@@ -0,0 +1,255 @@
+  settingsButton->setAutoRaise( true );
+  settingsButton->setToolTip( tr( "Settings" ) );
+  settingsButton->setMenu( QgisApp::instance()->gpsSettingsMenu() );
+  settingsButton->setPopupMode( QToolButton::InstantPopup );
+  settingsButton->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionOptions.svg" ) ) );
+  addWidget( settingsButton );
+
+  mRecenterAction->setEnabled( false );
+  mAddFeatureAction->setEnabled( false );
+  mAddTrackPointAction->setEnabled( false );
+  connect( mConnection, &QgsAppGpsConnection::statusChanged, this, [ = ]( Qgis::GpsConnectionStatus status )
+  {
+    switch ( status )
+    {
+      case Qgis::GpsConnectionStatus::Disconnected:
+        whileBlocking( mConnectAction )->setChecked( false );
+        mConnectAction->setText( tr( "Connect GPS" ) );
+        mConnectAction->setToolTip( tr( "Connect to GPS" ) );
+        mConnectAction->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/gpsicons/mIconGpsConnect.svg" ) ) );
+        mConnectAction->setEnabled( true );
+        mRecenterAction->setEnabled( false );
+        mAddFeatureAction->setEnabled( false );
+        mAddTrackPointAction->setEnabled( false );
+        break;
+      case Qgis::GpsConnectionStatus::Connecting:
+        whileBlocking( mConnectAction )->setChecked( true );
+        mConnectAction->setToolTip( tr( "Connecting to GPS" ) );
+        mConnectAction->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/gpsicons/mIconGpsConnect.svg" ) ) );
+        mConnectAction->setEnabled( false );
+        mRecenterAction->setEnabled( false );
+        mAddFeatureAction->setEnabled( false );
+        mAddTrackPointAction->setEnabled( false );
+        break;
+      case Qgis::GpsConnectionStatus::Connected:
+        whileBlocking( mConnectAction )->setChecked( true );
+        mConnectAction->setText( tr( "Disconnect GPS" ) );
+        mConnectAction->setToolTip( tr( "Disconnect from GPS" ) );
+        mConnectAction->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/gpsicons/mIconGpsDisconnect.svg" ) ) );
+        mConnectAction->setEnabled( true );
+        mRecenterAction->setEnabled( true );
+        mAddFeatureAction->setEnabled( true );
+        mAddTrackPointAction->setEnabled( mEnableAddVertexButton );
+        break;
+    }
+  } );
+
+  connect( QgisApp::instance(), &QgisApp::activeLayerChanged,
+           this, &QgsGpsToolBar::updateCloseFeatureButton );
+}
+
+void QgsGpsToolBar::setAddVertexButtonEnabled( bool enabled )
+{
+  mEnableAddVertexButton = enabled;
+  mAddTrackPointAction->setEnabled( mEnableAddVertexButton && mConnection->isConnected() );
+}
+
+void QgsGpsToolBar::updateLocationLabel( const QgsPoint &point )
+{
+  if ( point.isEmpty() )
+  {
+    mLocationLabel->clear();
+  }
+  else
+  {
+    const QString pos = QgsCoordinateUtils::formatCoordinateForProject( QgsProject::instance(), point, QgsCoordinateReferenceSystem(), 8 );
+    mLocationLabel->setText( pos );
```

Comment:
You know what would be good here? Have a tooltip that shows more details (e.g. horizontal/vertical accuracy etc.)

Reply:
There's a follow up planned here, where I'll give right-click options for what the label should show (including current track length)